### PR TITLE
avoid ftruncate twice in buffered io

### PR DESCRIPTION
**Pull request by lightmark**
in buffered io, the filesize_ is the real size.

### CODE REVIEW

File: util/io_posix.cc [hunk 1]
```diff
@@ -730,11 +730,13 @@ Status PosixWritableFile::PositionedAppend(const Slice& data, uint64_t offset) {
 
 Status PosixWritableFile::Truncate(uint64_t size) {
   Status s;
-  int r = ftruncate(fd_, size);
-  if (r < 0) {
-    s = IOError(filename_, errno);
-  } else {
-    filesize_ = size;
+  if (use_direct_io()) {
```
Comment (siying) [inline util/io_posix.cc]: I think this logic should happen in WritableFileWriter. It's more straight-forward if File::Truncate() always truncates.

Comment (lightmark) [inline util/io_posix.cc]: good idea

File: util/file_reader_writer.cc [hunk 2]
```diff
@@ -184,11 +184,14 @@ Status WritableFileWriter::Close() {
 
   s = Flush();  // flush cache to OS
 
+  Status interim;
```
Comment (siying) [inline util/file_reader_writer.cc]: Now this can happen in line 191?

File: util/file_reader_writer.cc [hunk 3]
```diff
@@ -184,11 +184,14 @@ Status WritableFileWriter::Close() {
...
   // In direct I/O mode we write whole pages so
   // we need to let the file know where data ends.
-  Status interim = writable_file_->Truncate(filesize_);
-  if (!interim.ok() && s.ok()) {
-    s = interim;
+  if (use_direct_io()) {
```
Comment (siying) [inline util/file_reader_writer.cc]: Please make sure it works on Windows.

File: util/file_reader_writer.cc (hunk 2 above)
Comment (lightmark) [inline util/file_reader_writer.cc]: line 198 also needs this

File: util/file_reader_writer.cc (hunk 3 above)
Comment (lightmark) [inline util/file_reader_writer.cc]: Yes. WinWritableFile::use_direct_io() exists already

**Pull request closed without merging**